src/objc/AudioPreprocessing.m

```diff
@@ -0,0 +1,365 @@
+    NSAssert(_maxFreq != 0, @"Max frequency cannot be 0.");
+    NSAssert(_sampleRate != 0, @"Sample rate cannot be 0.");
+    NSAssert(_numMelBands != 0, @"numMelBands cannot be 0.");
+    NSAssert(_numBins != 0, @"numBins cannot be 0.");
+
+    _filters = [[NSMutableArray alloc] initWithCapacity:numMelBands];
+
+    float melMin = hzToMel(_minFreq);
+    float melMax = hzToMel(_maxFreq);
+
+    // Get points equally spaced in the Mel Scale.
+    float melBandEdges[_numMelBands + 2];
+    float increment = (melMax - melMin) / (_numMelBands + 1);
+    vDSP_vramp(&melMin, &increment, melBandEdges, 1, _numMelBands + 2);
+
+    // Get points equally spaced in the Hz Scale.
+    float spectrogramBins[spectrumSize];
+    float minHz = 0.0;
+    float HzIncrement = (sampleRate/2) / (spectrumSize - 1);
+    vDSP_vramp(&minHz, &HzIncrement, spectrogramBins, 1, spectrumSize);
+
+    // Convert Hz to Mel
+    for(int i = 0; i < spectrumSize; i++) {
+        spectrogramBins[i] = hzToMel(spectrogramBins[i]);
+    }
+
+    // For each band edge, create a new filter (basically a float array that we vDSP_dot later)
+    for(size_t i = 0; i < _numMelBands; i++) {
+        // Grab the low, center, and bottom edges
+        float lower_edge = melBandEdges[i];
+        float center_edge = melBandEdges[i+1];
+        float upper_edge = melBandEdges[i+2];
+
+        // Compute lower and upper slopes
+        float* lowerBank = (float*)calloc(spectrumSize, sizeof(float));
+        float* upperBank = (float*)calloc(spectrumSize, sizeof(float));
+        float* filterBank = (float*)calloc(spectrumSize, sizeof(float));
+        float add = -lower_edge;
+        float div = center_edge - lower_edge;
+        vDSP_vsadd(spectrogramBins, 1, &add, lowerBank, 1, spectrumSize);
+        vDSP_vsdiv(lowerBank, 1, &div, lowerBank, 1, spectrumSize);
+
+        float negative_one = -1;
+        add = upper_edge;
+        div = upper_edge - center_edge ;
+        vDSP_vsmul(spectrogramBins, 1, &negative_one, upperBank, 1, spectrumSize);
+        vDSP_vsadd(upperBank, 1, &add, upperBank, 1, spectrumSize);
+        vDSP_vsdiv(upperBank, 1, &div, upperBank, 1, spectrumSize);
+
+        // find intersection
+        for(int k = 0; k < spectrumSize; k++) {
+            filterBank[k] = (k == 0) ? 0.0 : fmax(0.0, fmin(lowerBank[k], upperBank[k]));
+        }
+
+        _filterBanks[i] = calloc(spectrumSize, sizeof(float));
+        _filterBanks[i] = filterBank;
+    }
+
+    return self;
+}
+
+-(void)apply:(const float *)powerSpectrum
+      output:(float *)out {
+
+  float res;
+  for(size_t i = 0; i < _numMelBands; i++) {
+    vDSP_dotpr(powerSpectrum, 1, _filterBanks[i], 1, &res, spectrumSize);
+    out[i] = log(res + log_offset);
+  }
+}
+
+@end   // MelFrequencyFilterBank implementation
+
+
+
+@implementation TCSoundClassifierPreprocessing
+
+PowerSpectrum* _powerSpectrum;
+MelFrequencyFilterBank* _melFilterBank;
```

Comment:
Huh, I note that there aren't curly braces around these instance variables. Are those no longer needed these days? Or are these actually global variables?

Note that it's often cleaner to define internal ("private") variables as class extension properties, e.g.
```
@interface TCSoundClassifierPreprocessing ()
@property (nonatomic) PowerSpectrum *powerSpectrum;
@property (nonatomic) MelFrequenceFilterPank *melFilterBank;
@end
```
This will auto-synthesize underlying instance variables such as `_powerSpectrum` and allow you to use property syntax (e.g. `self.powerSpectrum`) outside of the initializer

Reply:
I don't know what's going on here either. I've added curly braces.